### PR TITLE
HAWQ-1100. Support for new filter string

### PR DESCRIPTION
**Pull request by kavinderd**
- constant data type is now of the form "c<DATATYPE>s<LENGTH>d<DATA>"

### CODE REVIEW

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 1]
```diff
@@ -279,25 +282,108 @@ int safeToInt(Long value) throws FilterStringSyntaxException {
...
+            return Integer.parseInt(digits);
+        } catch (NumberFormatException e) {
+            throw new FilterStringSyntaxException("invalid numeric argument " + digits);
+        }
+    }
+
+    private Object convertDataType(byte[] byteData, DataType dataType) throws Exception {
+        String data = new String(byteData);
+        try {
+            switch (dataType) {
```
Comment (sansanichfb) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: Should we also include TIME and BYTEA types?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 2]
```diff
@@ -165,17 +168,17 @@ public FilterParser(FilterBuilder eval) {
...
             throw new FilterStringSyntaxException("filter parsing ended with no result");
         }
 
-        while (index < filterString.length()) {
-            char op = filterString.charAt(index);
+        while (index < filterByteArr.length) {
+            char op = (char) filterByteArr[index];
             ++index; // skip op character
             switch (op) {
                 case 'a':
```
Comment (sansanichfb) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: Should we have all conventional characters used in protocol as constants or enum?
'a','c', 's','d','o','l'

File: pxf/pxf-api/src/test/java/org/apache/hawq/pxf/api/FilterParserTest.java [hunk 3]
```diff
@@ -143,32 +167,39 @@ public void parseNegativeBadOperation() {
...
         runParseNegative("filter with only column", filter, "filter parsing failed, missing operators?");
 
-        filter = "c1";
+        filter = "c20s1d1";
         runParseNegative("filter with only numeric const", filter, "filter parsing failed, missing operators?");
+    }
 
-        filter = "c\"something in the way\"";
-        runParseNegative("filter with only string const", filter, "filter parsing failed, missing operators?");
+    @Test
```
Comment (sansanichfb) [inline pxf/pxf-api/src/test/java/org/apache/hawq/pxf/api/FilterParserTest.java]: Add test case for negative value?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 4]
```diff
@@ -279,25 +282,108 @@ int safeToInt(Long value) throws FilterStringSyntaxException {
...
-                ? parseString()
-                : parseNumber();
-    }
+        DataType dataType = DataType.get(parseConstDataType());
+        if (dataType == DataType.UNSUPPORTED_TYPE) {
+            throw new FilterStringSyntaxException("invalid DataType OID at " + (index - 1));
+        }
+
+        int dataLength = parseDataLength();
+        if (dataLength < 1) {
```
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: How would an equality check for "" be handled ? What would the data length be ?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java (hunk 2 above)
Comment (kavinderd) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: Yeah, an enum is a good idea...I'll add it

File: pxf/pxf-hbase/src/main/java/org/apache/hawq/pxf/plugins/hbase/HBaseFilterBuilder.java [hunk 5]
```diff
@@ -229,8 +233,20 @@ private ByteArrayComparable getComparator(int type, Object data) throws Exceptio
...
-                result = new HBaseIntegerComparator((Long) data);
+                if (data instanceof Long) {
+                    result = new HBaseIntegerComparator((Long) data);
+                } else if (data instanceof Integer) {
+                    result = new HBaseIntegerComparator(((Integer) data).longValue());
+                } else {
+                    result = null;
+                }
+                break;
+            case FLOAT8:
```
Comment (kavinderd) [inline pxf/pxf-hbase/src/main/java/org/apache/hawq/pxf/plugins/hbase/HBaseFilterBuilder.java]: Good point

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 6]
```diff
@@ -279,25 +282,108 @@ int safeToInt(Long value) throws FilterStringSyntaxException {
...
+
+        try {
+            return Integer.parseInt(digits);
+        } catch (NumberFormatException e) {
+            throw new FilterStringSyntaxException("invalid numeric argument " + digits);
+        }
+    }
+
+    private Object convertDataType(byte[] byteData, DataType dataType) throws Exception {
+        String data = new String(byteData);
```
Comment (denalex) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: creating a string out of bytes assumes a specific charset. how will that be handled ?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 7]
```diff
@@ -318,17 +404,17 @@ private Long parseNumber() throws Exception {
     private String parseDigits() throws Exception {
         String result;
         int i = index;
-        int filterLength = filterString.length();
+        int filterLength = filterByteArr.length;
 
         // allow sign
         if (filterLength > 0) {
-            int chr = filterString.charAt(i);
+            int chr = (char) filterByteArr[i];
```
Comment (denalex) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: if we cast to (char), should the datatype be char as well, not int ? i know it works either way, but for consistency...

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java (hunk 6 above)
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: it would use the platform's default charset. @kavinderd is that the interned behavior ?
If not, use "UTF-8"

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 8]
```diff
@@ -356,8 +442,8 @@ private String parseString() throws Exception {
         int i;
 
         // starting from index + 1 to skip leading "
-        for (i = index + 1; i < filterString.length(); ++i) {
-            char chr = filterString.charAt(i);
+        for (i = index + 1; i < filterByteArr.length; ++i) {
+            char chr = (char) filterByteArr[i];
```
Comment (denalex) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: do we still have quotes around string values ? since we are passing length, it is no longer necessary

File: pxf/pxf-hbase/src/main/java/org/apache/hawq/pxf/plugins/hbase/HBaseFilterBuilder.java [hunk 9]
```diff
@@ -89,12 +90,15 @@ private boolean filterNotOpPresent(String filterString) {
...
+        if (filterString == null)
+            return null;
+
         // First check for NOT, HBase does not support this
         if (filterNotOpPresent(filterString))
             return null;
 
         FilterParser parser = new FilterParser(this);
-        Object result = parser.parse(filterString);
+        Object result = parser.parse(filterString.getBytes());
```
Comment (denalex) [inline pxf/pxf-hbase/src/main/java/org/apache/hawq/pxf/plugins/hbase/HBaseFilterBuilder.java]: this is confusing ... we already have a string here, then we convert it to byte[] to send to parser which converts each byte to char ... why ?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 10]
```diff
@@ -162,6 +165,10 @@ private boolean buildArgument(SearchArgument.Builder builder, Object filterObj)
         ColumnDescriptor filterColumn = inputData.getColumn(filterColumnIndex);
         String filterColumnName = filterColumn.columnName();
 
+        /* Need to convert java.sql.Date to Hive's DateWritable Format */
+        if (filterValue instanceof Date)
```
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Why does the type conversion only need to happen for Date and not for other types ?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveFilterBuilder.java [hunk 11]
```diff
@@ -61,8 +61,11 @@ public HiveFilterBuilder(InputData input) {
      *             filter or list of basic filters
      */
     public Object getFilterObject(String filterString) throws Exception {
+        if (filterString == null)
```
Comment (denalex) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveFilterBuilder.java]: style : curly brackets missing ?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 12]
```diff
@@ -279,25 +282,108 @@ int safeToInt(Long value) throws FilterStringSyntaxException {
...
+
+        if (((char) filterByteArr[index]) != 'd') {
+            throw new FilterStringSyntaxException("data delimiter 'd' expected at " + index);
+        }
+
+        index++;
 
-    private boolean senseString() {
-        return filterString.charAt(index) == '"';
+        Object data = convertDataType(Arrays.copyOfRange(filterByteArr, index, index+dataLength), dataType);
```
Comment (kavinderd) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: Good catch, I'll update the method

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java (hunk 4 above)
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: For the case when TEXT field is checked with = against "", bridge uses s0do5 in the expression string.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java (hunk 10 above)
Comment (kavinderd) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Hive doesn't support java.sql.Data so I needed to catch this explicitly and convert to DateWritable

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 13]
```diff
@@ -279,25 +296,110 @@ int safeToInt(Long value) throws FilterStringSyntaxException {
...
+
+        try {
+            return Integer.parseInt(digits);
+        } catch (NumberFormatException e) {
+            throw new FilterStringSyntaxException("invalid numeric argument " + digits);
+        }
+    }
+
+    private Object convertDataType(byte[] byteData, int start, int end, DataType dataType) throws Exception {
+        String data = new String(byteData, start, end-start);
```
Comment (denalex) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: didn't we want to explicitly specify UTF-8 here, since the data was re-encoded with UTF-8 after being parsed from the header ?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java [hunk 14]
```diff
@@ -30,26 +34,40 @@
  * interface with two pop-ed operands.
  * <br>
```
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java]: Can you add this 
The filter string is of the pattern `<attcode><attnum><constcode><constval><constsizecode><constsize><constdata><constvalue><opercode><opernum>`

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveFilterBuilder.java [hunk 15]
```diff
@@ -61,8 +61,11 @@ public HiveFilterBuilder(InputData input) {
...
+            return null;
+
         FilterParser parser = new FilterParser(this);
-        Object result = parser.parse(filterString);
+        Object result = parser.parse(filterString.getBytes());
```
Comment (denalex) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveFilterBuilder.java]: and here ?
